### PR TITLE
added concerns-spec

### PR DESCRIPTION
**Pull request by wynn987**
## Description
- Test for companies concern
- 155386546

### CODE REVIEW

File: spec/concerns/statistics/companies_spec.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  end
+  let(:valid_service_review) do
+    build(:service_review).attributes
+  end
+  let(:valid_company) do
+    build(:company, aggregate_score: 0).attributes
+  end
+
+  describe "reviews_count" do
+    context "no product or service" do
```
Comment (kklw) [inline spec/concerns/statistics/companies_spec.rb]: no product and service

File: app/models/concerns/statistics/companies.rb [hunk 2]
```diff
@@ -7,57 +7,70 @@ module Statistics::Companies
...
-      + self.services.reduce(0) { |accum, s| accum + s.reviews.count }
+      self.products.reduce(0) { |accum, pr| accum + pr.reviews.count } + self.services.reduce(0) { |accum, s| accum + s.reviews.count }
     end
 
     def strengths
-      strengths = Set.new()
-      strengths = get_product_reviews(strengths)
-      strengths = get_service_reviews(strengths)
-      strengths.to_a
+      product_strength_list = product_reviews
```
Comment (kklw) [inline app/models/concerns/statistics/companies.rb]: rename to set

File: app/models/concerns/statistics/companies.rb [hunk 3]
```diff
@@ -7,57 +7,70 @@ module Statistics::Companies
...
+        ((count * aggregate_score) - score)/(count - 1)
+      else
+        0
+      end
     end
 
     private
 
-    def get_product_reviews(set)
+    def product_reviews
```
Comment (kklw) [inline app/models/concerns/statistics/companies.rb]: merge with service_reviews function

File: app/models/concerns/statistics/companies.rb [hunk 4]
```diff
@@ -7,57 +7,70 @@ module Statistics::Companies
...
 
-    def get_service_reviews(set)
-      services.first(3).each do |s|
-        s.reviews.first(3).each do |r|
-          if set.length >= 5
-            return set.to_a[0, 5]
-          end
-          set.merge(r.strengths)
+    def service_reviews
+      service_set = Set.new
```
Comment (kklw) [inline app/models/concerns/statistics/companies.rb]: should be strength_set

File: spec/concerns/statistics/companies_spec.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      it "returns 0" do
+        @company = Company.create! valid_company
+        expect(@company.reviews_count).to eq(0)
+      end
+    end
+
+    context "no product" do
+      it "returns service count" do
+        @company = Company.create! valid_company
```
Comment (kklw) [inline spec/concerns/statistics/companies_spec.rb]: remove unnecessary @

File: app/models/concerns/statistics/companies.rb [hunk 6]
```diff
@@ -7,57 +7,61 @@ module Statistics::Companies
     # These refer to the aggregate number of reviews of a vendor company's products and services
     # (different from #reviews method on company, see models/company.rb)
     def reviews_count
-      self.products.reduce(0) { |accum, p| accum + p.reviews.count }
-      + self.services.reduce(0) { |accum, s| accum + s.reviews.count }
+      product_count = self.products.reduce(0) { |accum, pr| accum + pr.reviews.count }
```
Comment (kklw) [inline app/models/concerns/statistics/companies.rb]: Update single letter vars to a word. 
product
service
review

File: app/models/concerns/statistics/companies.rb [hunk 7]
```diff
@@ -7,57 +7,61 @@ module Statistics::Companies
...
-      strengths = get_product_reviews(strengths)
-      strengths = get_service_reviews(strengths)
-      strengths.to_a
+      product_strength_set = get_reviews(self.products)
+      service_strength_set = get_reviews(self.services)
+      whole_set = product_strength_set.merge(service_strength_set)
+      if whole_set.empty?
+        []
+      else
+        whole_set.first(6).to_a
```
[the remark this answers is not in the document]
Comment (kklw) [reply app/models/concerns/statistics/companies.rb]: or 
```
whole_set.empty? ? [] : whole_set.first(6).to_a
```

File: app/models/concerns/statistics/companies.rb [hunk 8]
```diff
@@ -7,57 +7,61 @@ module Statistics::Companies
...
+      end
     end
 
     def subtract_score(score)
       count = reviews_count
-      ((count * aggregate_score) - score)/(count - 1)
+      if count > 1
+        ((count * aggregate_score) - score)/(count - 1)
+      else
+        0
```
[the remark this answers is not in the document]
Comment (wynn987) [reply app/models/concerns/statistics/companies.rb]: yea because its used twice so dont compute twice

[the remark this answers is not in the document]
Comment (kklw) [reply app/models/concerns/statistics/companies.rb]: Ah i miss that. You can just try the `ZeroDivisionError`

**Pull request merged**